Training/interface/DataLoader_main.h

```diff
@@ -0,0 +1,770 @@
+    }
+
+    void Initialize(std::string file_name)
+    {
+        if(tauTuple)
+            throw std::runtime_error("DataLoader is already initialized.");
+        file = OpenRootFile(file_name);
+        tauTuple = std::make_shared<tau_tuple::TauTuple>(file.get(), true);
+    }
+
+    bool HasNext() {
+        return (current_entry + n_tau) < end_dataset;
+    }
+
+    std::shared_ptr<Data> LoadNext(){
+
+        if(!tauTuple)
+            throw std::runtime_error("DataLoader is not initialized.");
+
+        auto data = std::make_shared<Data>(n_tau, n_fe_tau,
+                                           n_inner_cells, n_outer_cells,
+                                           n_pf_el,
+                                           n_pf_mu,
+                                           n_pf_chHad,
+                                           n_pf_nHad,
+                                           n_pf_gamma,
+                                           n_electrons,
+                                           n_muons,
+                                           n_labels
+                                          );
+
+        const Long64_t end_entry = current_entry + n_tau;
+        size_t n_processed = 0, n_total = static_cast<size_t>(end_entry - start_dataset);
+        for(Long64_t tau_i = 0; tau_i < n_tau; ++tau_i, ++current_entry) {
+            tauTuple->GetEntry(current_entry);
+            const auto& tau = tauTuple->data();
+            if(parity == -1 || tau.evt % 2 == static_cast<unsigned>(parity)) {
+                FillTauBranches(tau, tau_i, data);
+                FillCellGrid(tau, tau_i, innerCellGridRef, data, true);
+                FillCellGrid(tau, tau_i, outerCellGridRef, data, false);
+            }
+        }
+
+        std::cout << "One batch is returned" << std::endl;
+        return data;
+    }
+
+  private:
+      static constexpr float pi = boost::math::constants::pi<float>();
+
+      template<typename Scalar>
+      static Scalar DeltaPhi(Scalar phi1, Scalar phi2)
+      {
+          static constexpr Scalar pi = boost::math::constants::pi<Scalar>();
+          Scalar dphi = phi1 - phi2;
+          if(dphi > pi)
+              dphi -= 2*pi;
+          else if(dphi <= -pi)
+              dphi += 2*pi;
+          return dphi;
+      }
+
+
+      void FillTauBranches(const Tau& tau, Long64_t tau_i, std::shared_ptr<Data>& data)
+      {
+        Long64_t start_array_index = tau_i * n_fe_tau;
+
+        // Filling Tau Branche
+        auto get_tau_branch = [&](TauFlat_f _fe) -> float& {
+            size_t _fe_ind = static_cast<size_t>(_fe);
+            size_t index = start_array_index + _fe_ind;
+            return data->x_tau.at(index);
+        };
+
+        // filling labels one_hot vector
+        Int_t tau_type = tau.tauType;
+        if(tau_type<=3) data->y_onehot[ tau_i * n_labels + tau_type ] = 1;
+        else if(tau_type>=6) data->y_onehot[ tau_i * n_labels + tau_type - 2 ] = 1;
```

Comment:
this way to convert tau_type to y_onehot looks error-prone. For the default training setup, we have only 4 tau_types, all other taus (e.g. from embedded) should be ignored: or by moving to the next tau (better) or by setting the weight to 0.